bpf/lib/ip_options.h

```diff
@@ -0,0 +1,204 @@
+
+/* Maximum number of IPv4 options to process. */
+#define MAX_IPV4_OPTS 3
+
+/* The minimum value for IHL which corresponds to a packet with no options.
+ *
+ * A standard IP packet header has 20 bytes and the IHL is the number of 32 byte
+ * words.
+ */
+#define IHL_WITH_NO_OPTS 5
+
+/* Signifies that options were parsed correctly but no trace ID was found. */
+#define TRACE_ID_NOT_FOUND 0
+
+/* Signifies a failure to determine the trace ID based on an unspecified error. */
+#define TRACE_ID_ERROR -1
+
+/* Signifies that the trace ID was found but it was invalid. */
+#define TRACE_ID_INVALID -2
+
+/* Signifies a failure to determine trace ID because the IP family was not found. */
+#define TRACE_ID_NO_FAMILY -3
+
+/* Signifies a failure to determine trace ID because the IP option length was
+ * not supported.
+ */
+#define TRACE_ID_UNSUPPORTED_LENGTH_ERROR -4
+
+/* Signifies trace points which are being ignored because they're in IPv6
+ * code and not supported yet.
+ */
+#define TRACE_ID_SKIP_IPV6 -100
+
+/* Converts a 64-bit integer from network byte order to host byte order. */
+static __always_inline __u64 ntohll(__u64 value)
+{
+	return ((value & 0x00000000000000FFULL) << 56) |
+		   ((value & 0x000000000000FF00ULL) << 40) |
+		   ((value & 0x0000000000FF0000ULL) << 24) |
+		   ((value & 0x00000000FF000000ULL) << 8)  |
+		   ((value & 0x000000FF00000000ULL) >> 8)  |
+		   ((value & 0x0000FF0000000000ULL) >> 24) |
+		   ((value & 0x00FF000000000000ULL) >> 40) |
+		   ((value & 0xFF00000000000000ULL) >> 56);
+}
+
+/* trace_id_from_ip4 parses the IP options and returns the trace ID.
+ *
+ * See trace_id_from_ctx for more info.
+ */
+static __always_inline int
+trace_id_from_ip4(struct __ctx_buff *ctx, __s64 *value, struct iphdr *ip4, __u8 trace_ip_opt_type)
+{
+	__u32 offset;
+	__u32 end;
+	int i;
+	__u8 opt_type;
+	__u8 optlen;
+
+	if (ip4->ihl <= IHL_WITH_NO_OPTS)
+		return TRACE_ID_NOT_FOUND;
+
+	offset = ETH_HLEN + sizeof(struct iphdr);
+	end = offset + (ip4->ihl << 2);
+
+#pragma unroll(MAX_IPV4_OPTS)
+	for (i = 0; i < MAX_IPV4_OPTS && offset < end; i++) {
+		/* We load the option header 1 field at a time since different types
+		 * have different formats.
+		 *
+		 * "Options 0 and 1 are exactly one octet which is their type field. All
+		 * other options have their one octet type field, followed by a one
+		 * octet length field, followed by length-2 octets of option data."
+		 *
+		 * Ref: https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml
+		 */
+
+		if (ctx_load_bytes(ctx, offset, &opt_type, 1) < 0)
+			return TRACE_ID_ERROR;
+
+		if (opt_type == IPOPT_END)
+			break;
+
+		if (opt_type == IPOPT_NOOP) {
+			offset++;
+			continue;
+		}
+
+		if (ctx_load_bytes(ctx, offset + 1, &optlen, 1) < 0)
+			return TRACE_ID_ERROR;
+
+		if (opt_type != trace_ip_opt_type) {
+			offset += optlen;
+			continue;
+		}
+
+		if (optlen != OPT16_LEN && optlen != OPT32_LEN && optlen != OPT64_LEN)
+			return TRACE_ID_INVALID;
```

Comment:
This looks unnecessary given the check and error with `TRACE_ID_UNSUPPORTED_LENGTH_ERROR` below.

Reply:
Yup, removing and updating tests.